### DISCUSSION TITLE
How inherited can you go?

### DISCUSSION THREAD

**Discussion by yjunechoe** [ggplot2-extenders/ggplot-extension-club#28] [Q&A]
I found a function I wrote a while back to see which ggprotos had the longest "inheritance chain" (is this the right description? idk OOP 😅) in a package.

```r
findmax <- function(pkg) {
  ns <- asNamespace(pkg)
  ggprotos <- grep("^(Stat|Geom)", ls(envir = ns), value = TRUE)
  ggprotos_nclass <- mget(ggprotos, envir = ns) |> 
    lapply(class) |> 
    lengths()
  ggprotos_nclass[ggprotos_nclass == max(ggprotos_nclass)]
}
findmax("ggplot2")
#> GeomAnnotationMap           GeomCol       GeomDensity 
#>                 6                 6                 6
```

Across extension packages, I recall ggforce being the "winner" on this metric. Still true in a small manually tested sample, but only by 1!

```r
findmax("ggforce")
#> GeomMarkEllipse    GeomMarkHull    GeomMarkRect 
#>               7               7               7
class(ggforce::GeomMarkEllipse)
#> [1] "GeomMarkEllipse" "GeomMarkCircle"  "GeomShape"       "GeomPolygon"    
#> [5] "Geom"            "ggproto"         "gg"
```

Please drop an example if you know of any crazy ones 😂

**Comment by teunbrand** [comment]
This isn't a 'look what deep classes I found', but it's a suggestion to measure inheritance differently.
Essentially, every time you `ggproto("x", Class)`, you create a child environment whose parent is accessible via the `super()` method.
Hence, we can measure the 'depth' of a class by how often we can access the parent without running out of parents.

``` r
class_depth <- function(x) {
  counter <- 0
  while (is.function(x$super)) {
    x <- x$super()
    counter <- counter + 1
  }
  counter
}
```
If we measure inheritance this way, we can see that `ggforce::GeomMarkCircle` is pretty deep, but also `GeomCol` is pretty deep.
``` r
library(ggplot2)
class_depth(Geom)
#> [1] 0
class_depth(GeomRect)
#> [1] 1
class_depth(GeomBar)
#> [1] 2
class_depth(GeomCol)
#> [1] 3
class_depth(ggproto(NULL, GeomCol))
#> [1] 4
class_depth(ggforce::GeomMarkCircle)
#> [1] 3
```
Another interesting tidbit that one can learn is that scales are cloned every time something is added to a plot via `ggplot_add()`.
This cloning adds another depth level to a scale.
In the example below, the two plots are almost identical, but due to where scales are declared, the scales have different depths.

```
p1 <- ggplot(mpg, aes(displ, hwy)) +
  scale_x_continuous() +
  geom_point() +
  theme_bw()

p2 <- ggplot(mpg, aes(displ, hwy)) +
  geom_point() +
  theme_bw() +
  scale_x_continuous()

class_depth(p1$scales$scales[[1]])
#> [1] 5
class_depth(p2$scales$scales[[1]])
#> [1] 3
```

<sup>Created on 2024-06-29 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Reply by yjunechoe** [reply]
Ah you're right! How could I have forgotten about `$super()`! 😅
